### PR TITLE
feat: introduce getSubTreeSpans function in clickhouse query builder & introduce smart trace detail algorithm

### PR DESCRIPTION
**Pull request by makeavish**

### CODE REVIEW

File: frontend/src/api/trace/getTraceItem.ts [hunk 1]
```diff
@@ -2,14 +2,14 @@ import axios from 'api';
...
+import { PayloadProps, GetTraceItemProps } from 'types/api/trace/getTraceItem';
 
 const getTraceItem = async (
-	props: Props,
+	props: GetTraceItemProps,
 ): Promise<SuccessResponse<PayloadProps> | ErrorResponse> => {
 	try {
 		const response = await axios.request<PayloadProps>({
-			url: `/traces/${props.id}`,
+			url: `/traces/${props.id}?spanId=${props.spanId || ''}&levelUp=${props.levelUp || ''}&levelDown=${props.levelDown || ''}`,
```
Comment (palashgdev) [inline frontend/src/api/trace/getTraceItem.ts:12]: please make util for this one ... if value is present then only append in the string else not

File: frontend/src/container/TraceDetail/index.tsx [hunk 2]
```diff
@@ -77,7 +78,7 @@ function TraceDetail({ response }: TraceDetailProps): JSX.Element {
 		if (activeSelectedId) {
 			history.replace({
 				pathname: history.location.pathname,
-				search: `?spanId=${activeSelectedId}`,
+				search: `?spanId=${activeSelectedId}&levelUp=${levelUp}&levelDown=${levelDown}`,
```
Comment (palashgdev) [inline frontend/src/container/TraceDetail/index.tsx:81]: same here

File: frontend/src/container/TraceDetail/index.tsx [hunk 3]
```diff
@@ -50,7 +50,8 @@ function TraceDetail({ response }: TraceDetailProps): JSX.Element {
 	// const [searchSpanString, setSearchSpanString] = useState('');
 	const [activeHoverId, setActiveHoverId] = useState<string>('');
 	const [activeSelectedId, setActiveSelectedId] = useState<string>(spanId || '');
-
+	const [levelUp] = useState<string | null>(urlQuery.get('levelUp'));
```
Comment (palashgdev) [inline frontend/src/container/TraceDetail/index.tsx:53]: if we are not updating the state probably there is no meaning of `useState`

File: frontend/src/pages/TraceDetail/index.tsx [hunk 4]
```diff
@@ -2,16 +2,21 @@ import { Typography } from 'antd';
...
+import useUrlQuery from 'hooks/useUrlQuery';
+import React, { useState } from 'react';
 import { useQuery } from 'react-query';
 import { useParams } from 'react-router-dom';
 import { Props as TraceDetailProps } from 'types/api/trace/getTraceItem';
 
 function TraceDetail(): JSX.Element {
 	const { id } = useParams<TraceDetailProps>();
+	const urlQuery = useUrlQuery();
+	const [spanId] = useState<string | null>(urlQuery.get('spanId'));
```
Comment (palashgdev) [inline frontend/src/pages/TraceDetail/index.tsx:14]: same here

File: frontend/src/types/api/trace/getTraceItem.ts [hunk 5]
```diff
@@ -1,7 +1,16 @@
+import internal from "stream";
```
Comment (palashgdev) [inline frontend/src/types/api/trace/getTraceItem.ts:1]: remove this one

**Review by palashgdev — CHANGES_REQUESTED** [review]

File: frontend/src/types/api/trace/getTraceItem.ts (hunk 5 above)
Comment (makeavish) [reply frontend/src/types/api/trace/getTraceItem.ts:1]: Done

File: frontend/src/container/TraceDetail/utils.ts [hunk 6]
```diff
@@ -98,3 +98,25 @@ export const getTreeLevelsCount = (tree: ITraceTree): number => {
...
+
+export const formGetTraceItemUrlParams = (
+	paramsMap: Map<string, string | null>,
+): string => {
+	let urlParams = '';
+	let index = 0;
+	paramsMap.forEach((value, key) => {
+		if (index === 0) {
+			if (value) {
+				urlParams = `?${key}=${value}`;
```
Comment (palashgdev) [inline frontend/src/container/TraceDetail/utils.ts:110]: can you please decodeURI and encodeURI value also ?

File: frontend/src/container/TraceDetail/utils.ts [hunk 7]
```diff
@@ -98,3 +98,25 @@ export const getTreeLevelsCount = (tree: ITraceTree): number => {
...
+			} else {
+				urlParams = `?${key}=`;
```
Comment (palashgdev) [inline frontend/src/container/TraceDetail/utils.ts:112]: handle with try and catch as there is `URI malformed`

for reference: `k8s_pod_name%20NIN%20(%27asd%27)asd3daafc+q32easdc%%%`

**Review by palashgdev — CHANGES_REQUESTED** [review]

File: frontend/src/api/trace/getTraceItem.ts [hunk 8]
```diff
@@ -1,15 +1,22 @@
...
 import { ErrorResponse, SuccessResponse } from 'types/api';
-import { PayloadProps, Props } from 'types/api/trace/getTraceItem';
+import { GetTraceItemProps, PayloadProps } from 'types/api/trace/getTraceItem';
 
 const getTraceItem = async (
-	props: Props,
+	props: GetTraceItemProps,
 ): Promise<SuccessResponse<PayloadProps> | ErrorResponse> => {
 	try {
+		const paramsMap = {
```
Comment (palashgdev) [inline frontend/src/api/trace/getTraceItem.ts:12]: there is no need to create new object called paramMap we can destructure

File: frontend/src/container/Trace/TraceTable/index.tsx [hunk 9]
```diff
@@ -44,7 +45,13 @@ function TraceTable(): JSX.Element {
...
 
 	const getLink = (record: TableType): string => {
-		return `${ROUTES.TRACE}/${record.traceID}?spanId=${record.spanID}`;
+		const paramsMap = {
+			spanId: record.spanID,
+			levelUp: 0,
+			levelDown: 0,
+		};
+		const urlParams = formUrlParams(paramsMap);
+		return `${ROUTES.TRACE}/${record.traceID}${urlParams}`;
```
Comment (palashgdev) [inline frontend/src/container/Trace/TraceTable/index.tsx:54]:
```suggestion
		return `${ROUTES.TRACE}/${record.traceID}${formUrlParams({
			spanId: record.spanID,
			levelUp: 0,
			levelDown: 0,
		})}`;
```

File: frontend/src/container/TraceDetail/index.tsx [hunk 10]
```diff
@@ -75,12 +82,18 @@ function TraceDetail({ response }: TraceDetailProps): JSX.Element {
...
+			const paramsMap = {
+				spanId: activeSelectedId,
+				levelUp,
+				levelDown,
+			};
+			const urlParams = formUrlParams(paramsMap);
 			history.replace({
 				pathname: history.location.pathname,
-				search: `?spanId=${activeSelectedId}`,
+				search: `${urlParams}`,
```
Comment (palashgdev) [inline frontend/src/container/TraceDetail/index.tsx:93]:
```suggestion
				search: `${formUrlParams({
				  spanId: activeSelectedId,
				  levelUp,
				  levelDown,
				})}`,
```

**Review by palashgdev — CHANGES_REQUESTED** [review]

**Review by palashgdev — DISMISSED** [review]
LGTM

**Review by palashgdev — DISMISSED** [review]

**Review by mindhash — APPROVED** [review]
LGTM. The metric range API should be extended in future releases.

**Pull request merged**